src/Models/NonhydrostaticModels/immersed_pressure_solver.jl

```diff
@@ -0,0 +1,152 @@
+using Oceananigans.ImmersedBoundaries: mask_immersed_field!
+
+using KernelAbstractions: @kernel, @index
+
+import Oceananigans.Solvers: precondition!
+import Oceananigans.Models.NonhydrostaticModels: solve_for_pressure!
+
+struct ImmersedPoissonSolver{R, G, S}
+    rhs :: R
+    grid :: G
+    pcg_solver :: S
+end
+
+@kernel function fft_preconditioner_right_hand_side!(preconditioner_rhs, rhs)
+    i, j, k = @index(Global, NTuple)
+    @inbounds preconditioner_rhs[i, j, k] = rhs[i, j, k]
+end
+
+# FFTBasedPoissonPreconditioner
+function precondition!(p, solver::FFTBasedPoissonSolver, rhs, args...)
+    grid = solver.grid
+    arch = architecture(grid)
+
+    launch!(arch, grid, :xyz,
+            fft_preconditioner_right_hand_side!,
+            solver.storage, rhs)
+
+    return solve!(p, solver, solver.storage)
+end
+
+function ImmersedPoissonSolver(grid;
+                               preconditioner = nothing,
+                               reltol = eps(eltype(grid)),
+                               abstol = 0,
+                               kw...)
+
+    if preconditioner == "FFT"
+        arch = architecture(grid)
+        preconditioner = PressureSolver(arch, grid)
+    end
+
+    rhs = CenterField(grid)
+
+    pcg_solver = PreconditionedConjugateGradientSolver(compute_laplacian!; reltol, abstol,
+                                                       preconditioner,
+                                                       template_field = rhs,
+                                                       kw...)
+
+    return ImmersedPoissonSolver(rhs, grid, pcg_solver)
+end
+
+@kernel function calculate_pressure_source_term!(rhs, grid, Δt, U★)
+    i, j, k = @index(Global, NTuple)
+    @inbounds rhs[i, j, k] = divᶜᶜᶜ(i, j, k, grid, U★.u, U★.v, U★.w) / Δt
+end
+
+@inline laplacianᶜᶜᶜ(i, j, k, grid, ϕ) = ∇²ᶜᶜᶜ(i, j, k, grid, ϕ)
+
+@kernel function laplacian!(∇²ϕ, grid, ϕ)
+    i, j, k = @index(Global, NTuple)
+    @inbounds ∇²ϕ[i, j, k] = laplacianᶜᶜᶜ(i, j, k, grid, ϕ)
+end
+
+function compute_laplacian!(∇²ϕ, ϕ)
+    grid = ϕ.grid
+    arch = architecture(grid)
+
+    fill_halo_regions!(ϕ)
+
+    launch!(arch, grid, :xyz, laplacian!, ∇²ϕ, grid, ϕ)
+
+    return nothing
+end
+
+function solve_for_pressure!(pressure, solver::ImmersedPoissonSolver, Δt, U★)
+    # TODO: Is this the right criteria?
+    min_Δt = eps(typeof(Δt))
+    Δt <= min_Δt && return pressure
+
+    rhs = solver.rhs
+    grid = solver.grid
+    arch = architecture(grid)
+
+    if grid isa ImmersedBoundaryGrid
+        underlying_grid = grid.underlying_grid
+    else
+        underlying_grid = grid
+    end
+
+    launch!(arch, grid, :xyz, calculate_pressure_source_term!,
+            rhs, underlying_grid, Δt, U★)
+
+    mask_immersed_field!(rhs, zero(grid))
+
+    # Solve pressure Pressure equation for pressure, given rhs
+    # @info "Δt before pressure solve: $(Δt)"
+    solve!(pressure, solver.pcg_solver, rhs)
+
+    return pressure
+end
+
+struct MITgcmPreconditioner end
```

Comment:
can we name this something bit more descriptive and less referential to MITgcm?

Reply:
Something like `SymmetricLocalPreconditioner`? I'm not entirely sure, could use some ideas too!

Reply:
Well I that sounds much better because at least it tries to describe something and not just refer to some other model.

I'm not sure what this preconditioner is doing so that's why I'm not in position to propose something. I don't even know what local refers to. But definitely a `SomethingThatDescribesMePreconditioner` is better than `SimilarToSomOtherModelOrSomebodyElsesPreconditioner`. That's what I wanted to convey with my comment.

Reply:
I believe this preconditioner is some kind of perturbation about a diagonal preconditioner, is that right? `DiagonallyDominantPreconditioner` or something. In fact shoudln't we also delete the wrong version that's already present @simone-silvestri ?

Reply:
Exactly, it should be the same one but corrected to be 3D and with the correct denominator (we should correct also the 2D version of it)

Reply:
Ok so maybe `DiagonallyDominantThreeDimensionalPreconditioner` and also `TwoDimensional`.